### ISSUE TITLE
Create function to query INBO gisservices

### ISSUE THREAD

**Issue by peterdesmet** [inbo/inborutils#19]
## Use case

for publishing occurrence records, we sometimes have the IFBL or UTM code and want to retrieve the centroid coordinates of those squares. Would be nice if can do this using the reference layers provided via http://gisservices.inbo.be/arcgis/rest/services (rather than some file).

Would like to have:

* a function to get the data from the datasets provided at http://gisservices.inbo.be/arcgis/rest/services, with stable names for the datasets
* a tutorial on how to link occurrences -> ref dataset -> calculate centroid coordinates.

---

Providing access to this official (?) layers would help R/Python users make better use of these public datasets. Note: right now, this is a *nice to have*, not a real need.

**Comment by ThierryO** [comment]
centroids from UTM should be easy. You only need to know the UTM zone, the grid cell size, some string manipulation and simple math. The conventions of the codes are easy.

I don't have much experience with IFBL but IMHO is can be as straightforward as the UTM solution.

In case a generic solutions required, then a small C++ library seems more relevant than a REST API.

**Comment by stijnvanhoey** [comment]
> In case a generic solutions required, then a small C++ library seems more relevant than a REST API.

??

The REST API, WMS, WFS are already there. What would be the functionality of the `small C++ library`?

**Comment by ThierryO** [comment]
You probably need as much code to talk to the API as you would need to do the calculation in R (or python). And the API requires to transfer the data while native code doesn't.
The C++ library is only relevant in case you want to have a single code base to be used in several languages (R, Python, ...).

**Comment by stijnvanhoey** [comment]
Maybe better to split the two questions from Peter.

1. Using existing **web services** from INBO to read GIS data with R

The code actually exists... It is [GDAL](http://www.gdal.org/)!  :-)

```r
library(gdalUtils)
library(rgdal)

# INBO ArcGis - WFS service of the UTM 10 hokken
inbo_dsn <- "WFS:http://gisservices.inbo.be/arcgis/services/UTMhokken/MapServer/WFSServer?request=GetCapabilities&service=WFS"
ogrinfo(dsn, layer = "utm10_ED50", so = TRUE)
```
output  `ogrinfo`:
```
[1] "Had to open data source read-only."                                                                                              
 [2] "INFO: Open of `WFS:http://gisservices.inbo.be/arcgis/services/UTMhokken/MapServer/WFSServer?request=GetCapabilities&service=WFS'"
 [3] "      using driver `WFS' successful."                                                                                            
 [4] "Metadata:"                                                                                                                       
 [5] "  TITLE=WFS"                                                                                                                     
 [6] ""                                                                                                                                
 [7] "Layer name: UTMhokken:utm10_ED50"                                                                                                
 [8] "Metadata:"                                                                                                                       
 [9] "  TITLE=utm10_ED50" 
...
```

Downloading data is provided by `ogr2ogr`:
```r
#-------------------------------------------------------------------------------
# Loading the entire layer:
#-------------------------------------------------------------------------------
# downloading a specific Layer (GDAL in the background, saves as file)
ogr2ogr(inbo_dsn, "utm10.shp", layer = "UTMhokken:utm10_ED50")
# read the result back into R
utm10 <- readOGR("./utm10.shp", stringsAsFactors = FALSE)
```
![image](https://user-images.githubusercontent.com/754862/33658254-5374797e-da7c-11e7-878b-94306cf09a96.png)

```r
#-------------------------------------------------------------------------------
# ONLY selecting a subset of the WFS
#-------------------------------------------------------------------------------
query <- "TAG IN ('ES38', 'ES41', 'ES42', 'ES43', 'ES44')"
ogr2ogr(inbo_dsn, "utm10_subset.shp", layer = "UTMhokken:utm10_ED50", where = query)
# read the result back into R
utm10_subset <- readOGR("./utm10_subset.shp", stringsAsFactors = FALSE)
```
![image](https://user-images.githubusercontent.com/754862/33658233-40741f3c-da7c-11e7-9fd1-c613e0c42c16.png)

Although, we could/should definitely make this easier for the INBO-user by masking part of the complexity provided by GDAL and probably check the functionalities of GDAL for windows users.

2. Calculating **centroids**

The code actually exists... It is [GEOS](http://geos.osgeo.org/doxygen/)!  :-)

```r
library(rgeos)
subset_centroids <- gCentroid(utm10_subset, byid = TRUE)
```

P.S. for a more visual overview of how R packages and underlying existing libraries relate: https://stijnvanhoey.github.io/course_gis_scripting/gis_r.html#35

**Comment by stijnvanhoey** [comment]
Considering the rather strange turned result of the UTM-grid in the first image, @jeroendereu, could you maybe provide some more info about the [WFS handling](http://gisservices.inbo.be/arcgis/services/UTMhokken/MapServer/WFSServer?request=GetCapabilities&service=WFS), more specifically how this could result in the strange behavior?

When I check the individual features, the `Name` versus `wfs:DefaultCRS` confuses me:
```
<wfs:FeatureType>
  <wfs:Name>UTMhokken:utm10_ETRS89</wfs:Name>
  <wfs:Title>utm10_ETRS89</wfs:Title>
  <wfs:DefaultCRS>urn:ogc:def:crs:EPSG::31370</wfs:DefaultCRS>
  <ows:WGS84BoundingBox>
    <ows:LowerCorner>2.408290763856107 50.60749852416224</ows:LowerCorner>
    <ows:UpperCorner>6.00489125203271 51.53394449522809</ows:UpperCorner>
  </ows:WGS84BoundingBox>
</wfs:FeatureType>
```
Should I consider the result as `ETRS89` or as `EPSG::31370`?

**Comment by stijnvanhoey** [comment]
Apparently (thanks @Roel) this is related to the type of identifier that is used to specify the projection which has an effect on the way the sequence of axis (`YX` vs `XY`) needs to be handled (and GDAL handles it). 

Information is there online, e.g. 
* https://github.com/deegree/deegree3/wiki/Axis-order-handling
* http://docs.geoserver.org/stable/en/user/services/wfs/basics.html#wfs-basics-axis
* http://webhelp.esri.com/arcgisserver/9.3/java/index.htm#wfs_service.htm#axis_order

tl;dr, we can rely on WFS 1.0.0 (add this to the URL) for which the EPSG:31370 notation is used, and this will by defeault be properly interpreted by GDAL...

```r
library(gdalUtils)
library(rgdal)

# INBO ArcGis - WFS service of the UTM 10 hokken
inbo_dsn <- "WFS:http://gisservices.inbo.be/arcgis/services/UTMhokken/MapServer/WFSServer?request=GetCapabilities&service=WFS&version=1.0.0"

# downloading a specific Layer (GDAL in the background, saves as file)
ogr2ogr(inbo_dsn, "utm10.shp", layer = "UTMhokken:utm10_ETRS89")

# read the result back into R
utm10 <- readOGR("./utm10.shp", stringsAsFactors = FALSE)
plot(utm10)
```
![image](https://user-images.githubusercontent.com/754862/33668294-a2305592-da9f-11e7-9ecb-389f6cf90fe0.png)

**Comment by stijnvanhoey** [comment]
@milotictanja, the `ggplot` version of the plot as well ;-)
(*remark*: this has indeed nothing to do with the YX/XY issue above, but just an illustration and reminder we should add the `prepare_spatial_for_ggplot` function to the package)

```
library(ggplot2)

#' Prepare a SpatialLinesDataFrame or SpatialPolygonsDataFrame for ggplot2
#' plotting
#'
#' @param spatial_df SpatialLinesDataFrame or SpatialPolygonsDataFrame
#' @return data.frame
prepare_spatial_for_ggplot <- function(spatial_df){
    
    df <- fortify(spatial_df)
    merge(df, as.data.frame(spatial_df))
}

utm_f <- prepare_spatial_for_ggplot(utm10)
ggplot(data = utm_f, aes(x = long, y = lat, group = group)) +
    geom_path() +
    coord_fixed()
```
![image](https://user-images.githubusercontent.com/754862/33669926-08383d24-daa4-11e7-950d-12934457848b.png)

**Comment by florisvdh** [comment]
I have two points to add regarding this interesting thread:

1. http://gisservices.inbo.be/arcgis/rest/services returns error 403: 'Services Directory has been disabled.' So is there another way of seeing which layers are available as WFS?
1. a potential problem for long-term reproducibility, depending on the use case, is that the data behind WFS-pointers may evolve. Is this the case for the INBO gisservices or are these WFSes used for fixed versions? An example of an evolving case (every two years) is the WFS for the Natura 2000 habitat map at Geopunt, which refers to the latest version.

**Comment by hansvancalster** [comment]
I think the questions raised in this issue have been partly addressed in the issue thread and in the tutorials repo. 

See https://inbo.github.io/tutorials/tutorials/spatial_wfs_services/ for ways to interact with WFS services and https://inbo.github.io/tutorials/tutorials/spatial_point_in_polygon/ for an example of occurrence data and square grid cells.

@peterdesmet is this issue still something you would like to see implemented as a dedicated function for the task?  I ask because the issue is old. If not I'd like to close this issue.

**Comment by peterdesmet** [comment]
No, the fact that there is a tutorial is great!
